File: lib/formats/arc/nsa_archive.rb
require 'lib/formats/gfx/spb_converter'
require 'lib/virtual_file'
require_relative 'nsa_archive/lzss_compressor'

# NSA archive
# Engine: Nscripter
# Extension: .nsa
# Known games:
# - Tsukihime
module NsaArchive
  module_function

  NO_COMPRESSION = 0
  SPB_COMPRESSION = 1
  LZSS_COMPRESSION = 2

  def add_cli_help(_arg_parser) end

  def parse_cli_options(_arg_parser, _options) end

  def lzss_compressor
    LzssCompressor.new(initial_dictionary_pos: 239, reuse_compressed: true)
  end

  class Unpacker
    def unpack(arc_file, output_files, _options)
      table = read_table(arc_file)
      read_contents(arc_file, table, output_files)
    end

    private

    def read_table(arc_file)
      num_files,
      offset_to_files = arc_file.read(6).unpack('S>L>')
      if offset_to_files > arc_file.size
        fail RecognitionError, 'Bad offset to files'
      end

      table = []
      num_files.times do
        e = {}

        e[:name] = arc_file.read_until_zero
        e[:compression_type],
        e[:origin],
        e[:size_compressed],
        e[:size_original] = arc_file.read(13).unpack('CL>L>L>')

        e[:origin] += offset_to_files
        table.push(e)

        if e[:origin] + e[:size_compressed] > arc_file.size
          fail RecognitionError, 'Bad offset to file'
        end
      end
      table
    end

    def read_contents(arc_file, table, output_files)
      table.each do |e|
        output_files.write do
          data = arc_file.peek(e[:origin]) do
            arc_file.read(e[:size_compressed])
          end

          file = VirtualFile.new(e[:name], data)
          decode!(file, e)

          unless file.data.length == e[:size_original]
            fail RecognitionError, 'Bad file size'
          end

          file
        end
      end
    end

    def decode!(file, table_entry)
      case table_entry[:compression_type]
      when SPB_COMPRESSION
        SpbConverter.decode!(file, {})
        table_entry[:size_original] = file.data.length
      when LZSS_COMPRESSION
        file.data = NsaArchive.lzss_compressor.decode(file.data)
      end
    end
  end

  class Packer
    def pack(arc_file, input_files, options)
      table_size = input_files.names.map { |n| n.length + 14 }.reduce(0, :+)
      offset_to_files = 6 + table_size
      arc_file.write([input_files.length, offset_to_files].pack('S>L>'))
      arc_file.write("\x00" * table_size)

      compression_type = options[:compression] || NO_COMPRESSION
      cur_data_origin = 0
      table_entries = []
      input_files.each do |file|
        data_size_original = file.data.length
        encode!(file, compression_type)
        data_size_compressed = file.data.length

        arc_file.write(file.data)

        table_entries.push([
          file.name,
          cur_data_origin,
          data_size_original,
          data_size_compressed])

        cur_data_origin += data_size_compressed
      end

      arc_file.seek(6)
      table_entries.each do |file_name, data_origin, orig_size, compressed_size|
        arc_file.write(file_name.gsub('/', '\\'))
        arc_file.write("\0")

        arc_file.write([
          compression_type,
          data_origin,
          compressed_size,
          orig_size].pack('CL>L>L>'))
      end
    end

    private

    def encode!(file, compression_type)
      case compression_type
      when SPB_COMPRESSION
        SpbConverter.encode!(file, {})
      when LZSS_COMPRESSION
        file.data = NsaArchive.lzss_compressor.encode(file.data)
      end
    end
  end
end

File: src/formats/arc/nsa_archive.cc
// NSA archive
//
// Company:   -
// Engine:    Nscripter
// Extension: .nsa
//
// Known games:
// - Tsukihime
// - Umineko no naku koro ni

#include "bit_reader.h"
#include "formats/arc/nsa_archive.h"

namespace
{
    typedef enum
    {
        COMPRESSION_NONE = 0,
        COMPRESSION_SPB = 1,
        COMPRESSION_LZSS = 2,
    } CompressionType;

    typedef struct
    {
        std::string name;
        CompressionType compression_type;
        size_t offset;
        size_t size_compressed;
        size_t size_original;
    } TableEntry;

    typedef std::vector<std::unique_ptr<TableEntry>> Table;

    Table read_table(IO &arc_io)
    {
        Table table;
        size_t file_count = arc_io.read_u16_be();
        size_t offset_to_files = arc_io.read_u32_be();
        if (offset_to_files > arc_io.size())
            throw std::runtime_error("Bad offset to files");

        for (size_t i = 0; i < file_count; i ++)
        {
            std::unique_ptr<TableEntry> entry(new TableEntry);
            entry->name = arc_io.read_until_zero();
            entry->compression_type
                = static_cast<CompressionType>(arc_io.read_u8());
            entry->offset = arc_io.read_u32_be();
            entry->size_compressed = arc_io.read_u32_be();
            entry->size_original = arc_io.read_u32_be();

            entry->offset += offset_to_files;

            if (entry->offset + entry->size_compressed > arc_io.size())
                throw std::runtime_error("Bad offset to file");

            table.push_back(std::move(entry));
        }
        return table;
    }

    void decompress_lzss(std::string &data, size_t size_original)
    {
        BufferedIO io(data);
        BitReader bit_reader(io);

        const size_t position_bits = 8;
        const size_t length_bits = 4;
        const size_t min_match_length = 2;
        const size_t initial_dictionary_pos = 239;
        const bool reuse_compressed = true;

        std::string output;
        output.reserve(size_original);
        size_t dictionary_size = 1 << position_bits;
        size_t dictionary_pos = initial_dictionary_pos;
        std::unique_ptr<unsigned char> dictionary(
            new unsigned char[dictionary_size]);

        unsigned char *dictionary_ptr = dictionary.get();

        size_t written = 0;
        while (written < size_original)
        {
            bool flag = bit_reader.get();

            if (flag)
            {
                unsigned char byte = bit_reader.get(8);
                output.push_back(byte);
                written ++;
                dictionary_ptr[dictionary_pos] = byte;
                dictionary_pos ++;
                dictionary_pos &= (dictionary_size - 1);
            }
            else
            {
                unsigned int pos = bit_reader.get(position_bits);
                unsigned int length = bit_reader.get(length_bits);
                length += min_match_length;
                for (size_t i = 0; i < length && written < size_original; i ++)
                {
                    unsigned char byte = dictionary_ptr[pos];
                    pos += 1;
                    pos &= (dictionary_size - 1);

                    if (reuse_compressed)
                    {
                        dictionary_ptr[dictionary_pos] = byte;
                        dictionary_pos ++;
                        dictionary_pos &= (dictionary_size - 1);
                    }
                    output.push_back(byte);
                    written ++;
                }
            }
        }

        data = output;
    }

    std::unique_ptr<VirtualFile> read_file(
        IO &arc_io, const TableEntry &table_entry)
    {
        std::unique_ptr<VirtualFile> file(new VirtualFile);

        file->name = table_entry.name;
        arc_io.seek(table_entry.offset);
        std::string data = arc_io.read(table_entry.size_compressed);

        switch (table_entry.compression_type)
        {
            case COMPRESSION_NONE:
                break;

            case COMPRESSION_LZSS:
                decompress_lzss(data, table_entry.size_original);
                break;

            case COMPRESSION_SPB:
                throw std::runtime_error("Not supported yet");
                break;
        }

        file->io.write(data);
        return file;
    }
}

struct NsaArchive::Internals
{
    // SPB converter will go here
};

NsaArchive::NsaArchive() : internals(new Internals)
{
}

NsaArchive::~NsaArchive()
{
}

void NsaArchive::unpack_internal(IO &arc_io, OutputFiles &output_files) const
{
    Table table = read_table(arc_io);
    for (size_t i = 0; i < table.size(); i ++)
    {
        output_files.save([&]() -> std::unique_ptr<VirtualFile>
        {
            return read_file(arc_io, *table[i]);
        });
    }
}
